core/controllers/reader_test.py

```diff
@@ -205,6 +248,103 @@ def test_huge_answers(self):
             'a' * 1000500, 'Sorry, nope, we didn\'t get it', '')
 
 
+class ReaderClassifyTests(test_utils.GenericTestBase):
+    """Test reader.classify using the sample explorations.
+
+    Since the end to end tests cover correct classification,
+    ReaderClassifyTests is only checking which of the hard/soft/classifier
+    rules triggered on an input.
+    """
+
+    def setUp(self):
+        super(ReaderClassifyTests, self).setUp()
+        self.init_classify_inputs('16')
+
+    def init_classify_inputs(self, exploration_id):
+        exp_services.delete_demo(exploration_id)
+        exp_services.load_demo(exploration_id)
```

Comment:
After looking through all of the tests, I'm slightly concerned about using a demo exploration here. These are unit tests, but the data they're operating on is not being tightly held. If the demo changes on the filesystem, it could change the behavior of these unit tests. I don't think that's very good. Unit tests should be atomic and as independent as possible from the rest of the system (including the demo explorations).

I suggest creating a new exploration and explicitly assembling the rules and test state to be exactly what these tests are supposed to operate upon. That will make the tests much clearer for anyone reading them, and also tighten their contract. Please let me know if you have any questions or concerns with this suggestion.